pulsar/producer_partition.go

```diff
@@ -481,10 +493,18 @@ func (p *partitionProducer) reconnectToBroker() {
 func (p *partitionProducer) runEventsLoop() {
 	for {
 		select {
-		case data := <-p.dataChan:
+		case data, ok := <-p.dataChan:
+			// when doClose() is call, p.dataChan will be closed, data will nil, we need to ignore it, or it will panic
+			if !ok {
+				continue
+			}
```

Comment:
When `doClose` is called. We could close this events loop, right? Could we just escape this event loop here?

Reply:
OK, I have updated this, we will exit the event loop when the chans are closed.